### ISSUE TITLE
Should Attributes be Stored in the AST?

### ISSUE THREAD

**Issue by InsertCreativityHere** [icerpc/slicec#430]
Currently, attributes are directly owned by the Slice entity they're applied to:
```
pub struct CustomType {
    pub parent: WeakPtr<Module>,       // Doesn't own it's parent module, since it only holds a pointer.
    pub attributes: Vec<Attribute>,    // Direct ownership, no pointers involved.
    pub doc_comment: DocComment,       // Same
    ... other fields
}
```
But it might simply our code if they were owned by the AST, and entities only held pointers to them.

----

Whether or not something belongs in the AST depends on: "**Do multiple things need to directly reference this data?**"
For instance, `CustomType`s need to be in the AST, because multiple things need to directly reference them:
- The module it's contained within needs to know it's contents.
- A data member might use this as it's data type

But `DocComment`s do not need to be in the AST.
Only one thing will ever care about a particular doc comment: the thing being documented.

----

Most of the time, `Attribute`s fall into the 2nd category: the only thing that cares about an attribute is what it's applied to.
But there is one exception to this rule: type-aliases.
```
typealias Foo = [attribute] bool;
struct S { a: Foo, b: Foo }
```
When we resolve the underlying type for the 2 data members, both of them will have the `[attribute]` applied to them.
In this case: **multiple things need to directly reference this attribute**. All of `Foo`, `a` and `b` need to reference it.

---

Our current solution to this is instead of them holding references to this single attribute, we actually clone a copy of the attribute everywhere where we need it. This is why `Attribute` is the _only_ grammar element that derives from `Clone`:
https://github.com/zeroc-ice/icerpc/blob/3e38d24071d887a2d8ef4f0d91b80b3f2f87c3de/src/grammar/elements/attribute.rs#L15-L16 and the machinery to make this struct clonable is weird looking.
It requires all the attribute kinds to also be cloneable, and we needed to add a `clone_kind` function to our trait, that all language mappings need to implement to create the clone, cast it, and move it onto the heap in a `Box`.
https://github.com/zeroc-ice/icerpc/blob/3e38d24071d887a2d8ef4f0d91b80b3f2f87c3de/src/grammar/elements/attribute.rs#L78-L79 https://github.com/zeroc-ice/icerpc/blob/3e38d24071d887a2d8ef4f0d91b80b3f2f87c3de/src/grammar/elements/attribute.rs#L93-L96 https://github.com/zeroc-ice/icerpc/blob/3e38d24071d887a2d8ef4f0d91b80b3f2f87c3de/src/grammar/elements/attribute.rs#L101-L105

TL;DR

We should look into whether or not storing Attributes in the AST make our code simpler and cleaner than allowing Slice entities to directly own them.
